Source/WebCore/dom/ContentVisibilityDocumentState.cpp

```diff
@@ -96,50 +94,82 @@ IntersectionObserver* ContentVisibilityDocumentState::intersectionObserver(Docum
     for (auto target : m_observer->observationTargets()) {
         if (target) {
-            auto oldRelevancy = target->contentRelevancy();
-            OptionSet<ContentRelevancy> newRelevancy;
-            if (oldRelevancy)
-                newRelevancy = *oldRelevancy;
-            auto setRelevancyValue = [&](ContentRelevancy reason, bool value) {
-                if (value)
-                    newRelevancy.add(reason);
-                else
-                    newRelevancy.remove(reason);
-            };
-            if (relevancyToCheck.contains(ContentRelevancy::OnScreen))
-                setRelevancyValue(ContentRelevancy::OnScreen, m_onScreenObservationTargets.contains(*target));
-
-            if (relevancyToCheck.contains(ContentRelevancy::Focused))
-                setRelevancyValue(ContentRelevancy::Focused, target->hasFocusWithin());
-
-            auto hasTopLayerinSubtree = [](const Element& target) {
-                for (auto& element : target.document().topLayerElements()) {
-                    if (element->isDescendantOf(target))
-                        return true;
-                }
-                return false;
-            };
-            if (relevancyToCheck.contains(ContentRelevancy::IsInTopLayer))
-                setRelevancyValue(ContentRelevancy::IsInTopLayer, hasTopLayerinSubtree(*target));
-
-            if (oldRelevancy && oldRelevancy == newRelevancy)
-                continue;
-            target->setContentRelevancy(newRelevancy);
-            target->invalidateStyle();
-            ContentVisibilityAutoStateChangeEvent::Init init;
-            init.skipped = newRelevancy.isEmpty();
-            target->queueTaskToDispatchEvent(TaskSource::DOMManipulation, ContentVisibilityAutoStateChangeEvent::create(eventNames().contentvisibilityautostatechangeEvent, init));
-            didUpdateAnyContentRelevancy = DidUpdateAnyContentRelevancy::Yes;
+            if (checkRelevancyOfContentVisibilityElement(*target, relevancyToCheck))
+                didUpdateAnyContentRelevancy = DidUpdateAnyContentRelevancy::Yes;
         }
     }
     return didUpdateAnyContentRelevancy;
 }
 
+HadInitialVisibleContentVisibilityDetermination ContentVisibilityDocumentState::determineInitialVisibleContentVisibility() const
+{
+    if (!m_observer)
+        return HadInitialVisibleContentVisibilityDetermination::No;
+    Vector<Ref<Element>> elementsToCheck;
+    for (auto target : m_observer->observationTargets()) {
+        if (target) {
+            bool checkForInitialDetermination = !m_viewportProximity.contains(*target) && !target->isRelevantToUser();
+            if (checkForInitialDetermination)
+                elementsToCheck.append(*target);
+        }
+    }
+    auto hadInitialVisibleContentVisibilityDetermination = HadInitialVisibleContentVisibilityDetermination::No;
+    if (!elementsToCheck.isEmpty()) {
+        elementsToCheck.first()->document().updateIntersectionObservations({ m_observer });
```

Comment:
Is it OK that this happens at a different point in the rendering update than other Intersection Observations? Could state have changed in between?

Reply:
The first usage of this intersection observer is special and meant to be same-frame as the element became c-v: auto, specifically to get correct behavior for the resize observers after that it acts just like any other intersection observer. Are you thinking of some specific state? The tests show up no problems.